src/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,371 @@
+        let mut quat = Quaternion::new(
+            0.0,
+            (1.0_f64 / 3.0_f64).sqrt() * 0.5, 
+            (1.0_f64 / 3.0_f64).sqrt() * -0.5, 
+            (1.0_f64 / 3.0_f64).sqrt() * 0.5
+        );
+        let expected_quat = Quaternion::new(
+            0.0,
+            (1.0_f64 / 3.0_f64).sqrt(), 
+            (1.0_f64 / 3.0_f64).sqrt() * -1.0, 
+            (1.0_f64 / 3.0_f64).sqrt()
+        );
+        quat.normalize();
+        assert!(approx_eq!(Quaternion, quat, expected_quat));
+    }
+
+    #[test]
+    fn is_normalized_is_correct() {
+        let not_normalized = Quaternion::new(
+            0.3,
+            (1.0_f64 / 3.0_f64).sqrt() * 0.5, 
+            (1.0_f64 / 3.0_f64).sqrt() * -0.5, 
+            (1.0_f64 / 3.0_f64).sqrt() * 0.5
+        );
+        assert!(!not_normalized.is_normalized());
+        let normalized = Quaternion::new(
+            0.0,
+            (1.0_f64 / 3.0_f64).sqrt(), 
+            (1.0_f64 / 3.0_f64).sqrt() * -1.0, 
+            (1.0_f64 / 3.0_f64).sqrt()
+        );
+        assert!(normalized.is_normalized());
+    }
+
+    #[test]
+    fn add_subtract_works() {
+        let quat1 = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        let quat2 = Quaternion::new(0.2, 0.3, 0.4, 0.5);
+        let expected_add = Quaternion::new(0.3, 0.5, 0.7, 0.9);
+        let expected_sub = Quaternion::new(-0.1, -0.1, -0.1, -0.1);
+        assert!(approx_eq!(Quaternion, quat1 + quat2, expected_add));
+        assert!(approx_eq!(Quaternion, quat1 - quat2, expected_sub));
+    }
+
+    #[test]
+    fn multiply_works() {
+        let quat1 = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        let quat2 = Quaternion::new(0.2, 0.3, 0.4, 0.5);
+        let expected_mul = Quaternion::new(-0.36, 0.06, 0.12, 0.12);
+        let expected_rev_mul = Quaternion::new(-0.36, 0.08, 0.08, 0.14);
+        assert!(approx_eq!(Quaternion, quat1 * quat2, expected_mul));
+        assert!(approx_eq!(Quaternion, quat2 * quat1, expected_rev_mul));
+    }
+
+    #[test]
+    fn scale_works() {
+        let mut quat = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        quat.scale(2.0);
+        let expected_quat = Quaternion::new(0.2, 0.4, 0.6, 0.8);
+        assert!(approx_eq!(Quaternion, quat, expected_quat));
+    }
+
+    #[test]
+    fn get_scaled_returns_scaled_quaternion() {
+        let quat = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        let quat_orig = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        let scaled_quat = quat.get_scaled(2.0);
+        let expected_quat = Quaternion::new(0.2, 0.4, 0.6, 0.8);
+        assert_eq!(quat, quat_orig);
+        assert!(approx_eq!(Quaternion, scaled_quat, expected_quat));
+    }
+
+    #[test]
+    fn conjugate_works() {
+        let quat = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        let expected_quat = Quaternion::new(0.1, -0.2, -0.3, -0.4);
+        assert!(approx_eq!(Quaternion, quat.conjugate(), expected_quat));
+    }
+
+    #[test]
+    fn quaternion_initializes_from_euler_angles() {
```

Comment:
[nit] Check near gimbal lock cases? https://danceswithcode.net/engineeringnotes/rotations_in_3d/demo3D/rotations_in_3d_tool.html

Reply:
I think one of these cases is near the gimbal lock? I'll double-check

Reply:
Gimbal lock is better tested in a set of rotations. I.e. after an axis aligns with another, do we lose a degree of freedom. It doesn't occur in one rotation. If out of scope for this ad you're limiting pitch in tait-bryan to pi/2, test that instead.